File: packages/ai/src/resilience.ts
import type { ModelRole } from '@konusbitr/shared';

/**
 * Retries, backoff and a circuit breaker, around one model call.
 *
 * Three failure shapes, three different right answers.
 *
 * A 429 or a 503 is the provider asking for a moment: retry with exponential
 * backoff and jitter. Jitter is not decoration — a worker embedding a
 * 500-page document in batches will hit a rate limit on many batches at once,
 * and a fixed backoff synchronises them into a thundering herd that hits the
 * same limit again in lockstep.
 *
 * A 400 or a 401 is the request or the credential being wrong: retrying is
 * pure latency, so it fails immediately and says which.
 *
 * A provider that is *down* is the case a breaker exists for. Without one, a
 * dead endpoint turns every job into `max_retries × timeout` of waiting before
 * it fails, and a queue of a hundred documents spends an hour discovering the
 * same outage a hundred times. After `MODEL_BREAKER_FAILURES` consecutive
 * failures the circuit opens and calls fail instantly for the cooldown, then
 * one call is let through to find out whether it is back.
 */

/** A model call that failed in a way worth naming. */
export class ModelCallError extends Error {
  override readonly name = 'ModelCallError';

  constructor(
    message: string,
    readonly options: {
      readonly role: ModelRole;
      readonly status?: number;
      readonly retryable: boolean;
      readonly cause?: unknown;
    },
  ) {
    super(message, { cause: options.cause });
  }

  get retryable(): boolean {
    return this.options.retryable;
  }
}

/** Raised while a role's circuit is open. Never retried: that is the point. */
export class CircuitOpenError extends Error {
  override readonly name = 'CircuitOpenError';

  constructor(
    readonly role: ModelRole,
    readonly reopensInMs: number,
  ) {
    super(
      `The ${role} provider has failed repeatedly; calls are being refused for ` +
        `another ${Math.ceil(reopensInMs / 1000)}s. Fix the provider rather than ` +
        'waiting out one timeout per job.',
    );
  }
}

/**
 * Status codes worth trying again.
 *
 * 408 and 409 are in because some OpenAI-compatible servers (vLLM under load,
 * Ollama while a model is still loading) use them for "busy, come back", and
 * 499 because a proxy in front of a slow local model can close the connection
 * before the model has finished loading on first use.
 */
const RETRYABLE_STATUSES = new Set([408, 409, 425, 429, 499, 500, 502, 503, 504]);

export function isRetryableStatus(status: number): boolean {
  return RETRYABLE_STATUSES.has(status);
}

export type BreakerOptions = {
  /** Consecutive failures that open the circuit. */
  failures: number;
  /** How long it stays open. */
  cooldownMs: number;
  /** Injected in tests; `Date.now` in production. */
  now?: () => number;
};

/**
 * One breaker per role.
 *
 * Per role rather than per process because the roles fail independently: a
 * deployment with a cloud chat model and local embeddings has two providers,
 * and an OpenAI outage must not stop the worker from embedding.
 */
export class CircuitBreaker {
  private consecutiveFailures = 0;
  private openedAt: number | undefined;

  constructor(private readonly options: BreakerOptions) {}

  private get now(): number {
    return (this.options.now ?? Date.now)();
  }

  /** Raise if the circuit is open and its cooldown has not elapsed. */
  assertClosed(role: ModelRole): void {
    if (this.openedAt === undefined) return;

    const elapsed = this.now - this.openedAt;
    if (elapsed < this.options.cooldownMs) {
      throw new CircuitOpenError(role, this.options.cooldownMs - elapsed);
    }

    // Half-open: let exactly one call through. The counter is left where it is,
    // so a single further failure re-opens immediately rather than granting a
    // fresh budget of attempts to a provider that is still down.
    this.openedAt = undefined;
  }

  recordSuccess(): void {
    this.consecutiveFailures = 0;
    this.openedAt = undefined;
  }

  recordFailure(): void {
    this.consecutiveFailures += 1;
    if (this.consecutiveFailures >= this.options.failures) {
      this.openedAt = this.now;
    }
  }

  /** For `/health`, and for the tests. */
  get state(): 'closed' | 'open' {
    if (this.openedAt === undefined) return 'closed';
    return this.now - this.openedAt < this.options.cooldownMs ? 'open' : 'closed';
  }
}

export type RetryOptions = {
  role: ModelRole;
  /** Attempts including the first. */
  attempts: number;
  breaker?: CircuitBreaker;
  /** Base delay; attempt *n* waits `base * 2 ** (n - 1)` plus jitter. */
  baseDelayMs?: number;
  /** Injected in tests so a retry test does not actually sleep. */
  sleep?: (ms: number) => Promise<void>;
  /** Injected in tests; `Math.random` in production. */
  random?: () => number;
};

const DEFAULT_BASE_DELAY_MS = 500;

function delayFor(attempt: number, baseMs: number, random: () => number): number {
  const exponential = baseMs * 2 ** (attempt - 1);
  // Full jitter: a uniform draw from [0, exponential] rather than
  // exponential ± a little. It is what keeps a batch of simultaneous rate
  // limits from retrying in lockstep.
  return Math.round(exponential * random());
}

/**
 * Run a model call with retries, backoff and the role's breaker.
 *
 * The breaker is consulted before the first attempt and updated after the last,
 * not per attempt: three attempts against one dead provider are one failure of
 * that provider, and counting them separately would open the circuit on the
 * first job rather than the fifth.
 */
export async function withResilience<T>(
  options: RetryOptions,
  call: (attempt: number) => Promise<T>,
): Promise<T> {
  const {
    role,
    attempts,
    breaker,
    baseDelayMs = DEFAULT_BASE_DELAY_MS,
    sleep = defaultSleep,
    random = Math.random,
  } = options;

  breaker?.assertClosed(role);

  let lastError: unknown;
  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    try {
      const result = await call(attempt);
      breaker?.recordSuccess();
      return result;
    } catch (error) {
      lastError = error;

      const retryable = error instanceof ModelCallError ? error.retryable : true;
      if (!retryable) {
        // A bad request or a rejected credential is not evidence that the
        // provider is unhealthy, so it must not count towards the breaker:
        // otherwise five malformed calls would take a working provider offline.
        throw error;
      }

      if (attempt === attempts) break;
      await sleep(delayFor(attempt, baseDelayMs, random));
    }
  }

  breaker?.recordFailure();
  throw lastError;
}

function defaultSleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: services/worker/src/konusbitr_worker/ai/resilience.py
"""Retries, backoff and a circuit breaker, around one model call.

Three failure shapes, three different right answers.

A 429 or a 503 is the provider asking for a moment: retry with exponential
backoff and **full jitter**. The jitter is not decoration. A worker embedding a
500-page document sends batches back to back, hits a rate limit on several of
them at once, and a fixed backoff synchronises those into a herd that hits the
same limit again in lockstep.

A 400 or a 401 is the request or the credential being wrong. Retrying is pure
latency, so it fails immediately, and — importantly — it does **not** count
towards the breaker: five malformed calls must not take a healthy provider
offline.

A provider that is genuinely down is why the breaker exists. Without one, a dead
endpoint costs every job ``attempts x timeout`` before failing, and a queue of a
hundred documents spends an hour discovering the same outage a hundred times.
"""

from __future__ import annotations

import asyncio
import random
import time
from collections.abc import Awaitable, Callable

from konusbitr_worker.log import get_logger
from konusbitr_worker.settings import ModelRole

__all__ = [
    "CircuitBreaker",
    "CircuitOpenError",
    "ModelCallError",
    "is_retryable_status",
    "with_resilience",
]

logger = get_logger("konusbitr.worker.ai.resilience")

#: Status codes worth trying again.
#:
#: 408, 409 and 425 are in because some OpenAI-compatible servers use them for
#: "busy, come back" — vLLM under load, and Ollama while a model is still being
#: loaded into memory, which on a first call can be a minute.
_RETRYABLE_STATUSES = frozenset({408, 409, 425, 429, 499, 500, 502, 503, 504})

_DEFAULT_BASE_DELAY_SECONDS = 0.5


def is_retryable_status(status: int) -> bool:
    return status in _RETRYABLE_STATUSES


class ModelCallError(RuntimeError):
    """A model call that failed in a way worth naming."""

    def __init__(
        self,
        message: str,
        *,
        role: ModelRole,
        retryable: bool,
        status: int | None = None,
    ) -> None:
        super().__init__(message)
        self.role = role
        self.retryable = retryable
        self.status = status


class CircuitOpenError(RuntimeError):
    """Raised while a role's circuit is open. Never retried: that is the point."""

    def __init__(self, role: ModelRole, reopens_in: float) -> None:
        super().__init__(
            f"The {role} provider has failed repeatedly; calls are being refused for "
            f"another {reopens_in:.0f}s. Fix the provider rather than waiting out one "
            "timeout per job."
        )
        self.role = role
        self.reopens_in = reopens_in


class CircuitBreaker:
    """One breaker per role.

    Per role rather than per process, because roles fail independently: a
    deployment with a cloud chat model and local embeddings has two providers,
    and an OpenAI outage must not stop the worker embedding.
    """

    def __init__(
        self,
        *,
        failures: int,
        cooldown_seconds: float,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        self._threshold = failures
        self._cooldown = cooldown_seconds
        self._clock = clock
        self._consecutive_failures = 0
        self._opened_at: float | None = None

    def assert_closed(self, role: ModelRole) -> None:
        if self._opened_at is None:
            return

        elapsed = self._clock() - self._opened_at
        if elapsed < self._cooldown:
            raise CircuitOpenError(role, self._cooldown - elapsed)

        # Half-open: let exactly one call through. The failure counter is left
        # where it is, so a single further failure re-opens the circuit
        # immediately rather than granting a fresh budget of attempts to a
        # provider that is still down.
        self._opened_at = None

    def record_success(self) -> None:
        self._consecutive_failures = 0
        self._opened_at = None

    def record_failure(self) -> None:
        self._consecutive_failures += 1
        if self._consecutive_failures >= self._threshold:
            self._opened_at = self._clock()
            logger.warning("model circuit opened", extra={"failures": self._consecutive_failures})

    @property
    def state(self) -> str:
        """``closed`` or ``open``. Reported by ``/health``, asserted by the tests."""
        if self._opened_at is None:
            return "closed"
        return "open" if self._clock() - self._opened_at < self._cooldown else "closed"


async def with_resilience[T](
    call: Callable[[int], Awaitable[T]],
    *,
    role: ModelRole,
    attempts: int,
    breaker: CircuitBreaker | None = None,
    base_delay_seconds: float = _DEFAULT_BASE_DELAY_SECONDS,
    sleep: Callable[[float], Awaitable[None]] = asyncio.sleep,
    jitter: Callable[[], float] = random.random,
) -> T:
    """Run a model call with retries, backoff and the role's breaker.

    The breaker is consulted before the first attempt and updated after the
    last, not per attempt: three attempts against one dead provider are one
    failure *of that provider*, and counting them separately would open the
    circuit on the first job rather than the fifth.
    """
    if breaker is not None:
        breaker.assert_closed(role)

    last_error: BaseException | None = None
    for attempt in range(1, attempts + 1):
        try:
            result = await call(attempt)
        except ModelCallError as error:
            last_error = error
            if not error.retryable:
                # Not evidence that the provider is unhealthy, so it must not
                # count towards the breaker.
                raise
        except Exception as error:
            # An unrecognised exception is treated as retryable, for the same
            # reason `internal` is a retryable job error code: wrongly deciding
            # something is permanent costs somebody their document.
            last_error = error
        else:
            if breaker is not None:
                breaker.record_success()
            return result

        if attempt == attempts:
            break
        # Full jitter: a uniform draw from [0, exponential] rather than the
        # exponential plus a little. It is what breaks the lockstep.
        await sleep(base_delay_seconds * 2 ** (attempt - 1) * jitter())

    if breaker is not None:
        breaker.record_failure()
    if last_error is None:  # pragma: no cover - the loop cannot exit without one
        raise RuntimeError("the retry loop ended with neither a result nor an error")
    raise last_error
